File: src/main/java/com/mobilesolutionworks/android/zero/EndPoint.java
package com.mobilesolutionworks.android.zero;

import android.content.Context;
import android.support.annotation.NonNull;

/**
 * Created by yunarta on 2/5/17.
 */

public abstract class EndPoint<T extends EndPoint.Contract> {

    private Struts.BackEnd struts;

    protected abstract T getInterface();

    protected void setup(@NonNull Struts.BackEnd struts) {
        this.struts = struts;
    }

    protected Context context() {
        return struts.context();
    }

    protected Struts.BackEnd struts() {
        return struts;
    }

    protected abstract void onStart();

    public interface Contract {
    }
}

File: src/main/kotlin/com/mobilesolutionworks/android/struts/EndPoint.kt
package com.mobilesolutionworks.android.struts

import android.util.Log

/**
 * Created by yunarta on 13/5/17.
 */

abstract class EndPoint<out T : EndPoint.Contract> {

    interface Contract

    internal var _struts: Struts.BackEnd<Scheduler.BackEnd>? = null
    protected val struts: Struts.BackEnd<Scheduler.BackEnd>?
        get() = _struts

    protected abstract val contract: T
    internal val _contract: T
        get() = contract

    internal fun setup(struts: Struts.BackEnd<Scheduler.BackEnd>) {
        this._struts = struts
    }

    protected val context: android.content.Context?
        get() = _struts?.context

    internal fun dispatchCreate(struts: Struts.Mutable) {
        onCreate(struts)
    }

    protected abstract fun onCreate(struts: Struts.Mutable)

    internal fun dispatchStart() {
        onStart()
    }

    protected abstract fun onStart()
}
